UI/MapWnd.cpp

```diff
@@ -2778,57 +2781,147 @@ void MapWnd::ClearSystemRenderingBuffers() {
-    if (start_sys_it==end_sys_it)
-        return path;
-
-    for (std::set<int>::const_iterator sysIt = resGroup.begin(); sysIt!=resGroup.end(); sysIt++)
-        ancestor[*sysIt] = -1;
-    ancestor[start_sys_it] = start_sys_it;
-    tryNext.push_back(start_sys_it);
-
-    while (!tryNext.empty() ) {
-        int sysID = tryNext.front();
-
-        TemporaryPtr<const System> system = objMap.Object<const System>(sysID);
-        if (!system)
-            continue;
-        const std::map<int, bool>& lanes = system->StarlanesWormholes();
-        for (std::map<int, bool>::const_iterator laneIt = lanes.begin();
-             laneIt != lanes.end(); ++laneIt)
-        {
-            int newSys = laneIt->first;
-            std::pair<int, int> lane_forward = std::make_pair(sysID, newSys);
-            std::pair<int, int> lane_backward = std::make_pair(newSys, sysID);
-            // see if this lane exists in this empire's supply propegation lanes set.  either direction accepted. if not, skip this lane
-            if (supplylanes.find(lane_forward) == supplylanes.end() && supplylanes.find(lane_backward) == supplylanes.end())
-                continue;
-            if (!laneIt->second && ( ancestor[newSys] == -1 )) { //is a starlane, and not yet visited newSys //TODO: should allow wormholes here?
-                ancestor[newSys] = sysID;
-                if (newSys==end_sys_it) {
-                    int iSys = newSys;
-                    while ((ancestor[iSys] !=-1)&&( ancestor[iSys] != iSys )) {
-                        path.push_back(iSys);
-                        iSys = ancestor[iSys];
+namespace {
+    // SupplyLaneMap map keyed by system containing all system
+    // corresponding to valid supply lane destinations
+    typedef boost::unordered_multimap<int,int> SupplyLaneMMap;
+
+    // AncestorInfo stores the \p ids of systems one back on the path.
+    struct AncestorInfo {
+        AncestorInfo(int a, int o) : ids(1, a), single_origin(o) {}
+        std::vector<int> ids;  // previous systems
+        // originating terminal or nothing for connected resources
+        boost::optional<int> single_origin;
+    };
+
+    struct PrevCurrInfo {
+        PrevCurrInfo(int p, int n, int o) : prev(p), curr(n), origin(o) {}
+        int prev, curr, origin;
+    };
+
+    /**
+       Returns either:
+       a) a \p good_path between from any \p terminal_points to all
+       reachable \p terminal_points along the \p supply_lanes.
+       b) empty indicating no more work to do.
+
+     */
+    void GetPathsThroughSupplyLanes(
+        boost::unordered_set<int> & good_path,
+        const boost::unordered_set<int> & terminal_points,
+        const SupplyLaneMMap& supply_lanes)
+    {
+        good_path.clear();
+
+        // No terminal points, so all paths lead nowhere.
+        if (terminal_points.empty())
+            return;
+
+        boost::unordered_map<int, AncestorInfo> ancestors;
+        std::deque<PrevCurrInfo> tryNext;
```

Comment:
`tryNext` was a variable name from the original code that survived the refactoring.  Goodbye `tryNext`.